Adding React state to markdown parser and save functinality

farhaanbukhsh opened the pull request:
This PR helps to convert the react state to markdown and send the data to edx-platform. This is required to show or have a markdown format of the  Problems.

~~**Discussions**: Link to any public dicussions about this PR or the design/architecture. Otherwise omit this.~~

~~**Dependencies**: None~~

~~**Screenshots**: Always include screenshots if there is any change to the UI.~~

~~**Sandbox URL**: TBD - sandbox is being provisioned.~~

~~**Merge deadline**: "None" if there's no rush, "ASAP" if it's critical, or provide a specific date if there is one.~~

**Testing instructions**:

1. Set up the devstack as mentioned [here](https://github.com/edx/frontend-lib-content-components/pull/110#issue-1360810689)
2. There are tests written for the react state parser which you can have a look and run using `npm test -- src/editors/containers/ProblemEditor/data/ReactStateParser.test.js --collectCoverage=false`
3. From studio if we are entering to the MFE and if any change is made and saved, this just reflects in the markdown and not in OLX so xblock preview doesn't have any change.
4. Only when you disable the flags and click edit you can see the change in the markdown editor.
5. You can also put a `console.log(content)` [here](https://github.com/open-craft/frontend-lib-content-components/pull/2/files#diff-c2d3d9d3c28ca1b4aa77706aa0aefb06b4d2f0facbb3a90e96a67963ed656eb3R122) for Gallery and similar place in api.js for live, you need to do a persist log here.

**Author notes and concerns**:

1. OLX has to be updated which will be done in the next tickets
2. We have still to find out how to parse and make the difference between question, description and label.

**Reviewers**
- [ ] @navinkarkera

Fixes: https://tasks.opencraft.com/browse/BB-6668
Co-authored-by: <>
Private-ref: <>

farhaanbukhsh committed:
feat: Implementation of react state markdown parser

This helps to convert the react state to markdown,
changes also have some cleaning up of the code and
implementation of save functionality.

Signed-off-by: Farhaan Bukhsh <farhaan@opencraft.com>

navinkarkera commented:
Initial review:
![Screenshot_20220915_133429](https://user-images.githubusercontent.com/10894099/190349990-2945f80e-2c01-4c3d-a07e-7ea86a7b3dbf.png)

farhaanbukhsh commented:
> Initial review: ![Screenshot_20220915_133429](https://user-images.githubusercontent.com/10894099/190349990-2945f80e-2c01-4c3d-a07e-7ea86a7b3dbf.png)

I have already added https://github.com/open-craft/frontend-lib-content-components/pull/2/files#diff-6a1fb3ea3560ddd90d64c5a7a704221c15242033e7234accc61ba1202e8c11a4R178

a TODO to support selected and unselected feedback, this is the markdown parser from capa problem visual editor, although we should fix it here, but a thought of using OLX instead of Markdown to parse to the editor makes more sense now, so this markdown parser can entirely be depricated.

This might be out of the scope of the current ticket as well.

navinkarkera commented on src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx:
```diff
@@ -1,14 +1,38 @@
       <div>
         <h1>
-          QuestionWidget
+          Question
```
Nit: missing internalization.

farhaanbukhsh replied:
Done

navinkarkera commented on src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx:
```diff
@@ -1,14 +1,38 @@
+      let content = editorRef.current.getContent({format: 'text'});
+      dispatch(actions.problem.updateQuestion(content));
+    }
+  }
```
Nit: can be replaced with `mapStateToProps` and `mapDispatchToProps`, https://github.com/open-craft/frontend-lib-content-components/blob/ce12d13552b5023afc7f0c8d7776bef1c65a2068/src/editors/containers/ProblemEditor/components/EditProblemView/AnswerWidget/AnswersContainer.jsx#L49-L55

farhaanbukhsh replied:
Done

navinkarkera commented on src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx:
```diff
@@ -6,18 +6,22 @@ import QuestionWidget from './QuestionWidget';
 export const EditProblemView = () => {
   const problemType = useSelector(selectors.problem.problemType);
-  const question = useSelector(selectors.problem.question);
+  const problemState = useSelector(selectors.problem.completeState);
```
Nit: same here, can be replaced with `mapStateToProps`

farhaanbukhsh replied:
thanks :)

navinkarkera reviewed (commented):
Some small suggestions. Feel free to ignore them if you feel it is not important

navinkarkera commented on src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx:
```diff
@@ -5,21 +5,31 @@ import SettingsWidget from './SettingsWidget';
+  problemState: selectors.problem.completeState(state),
+})
+
+export default injectIntl(connect(mapStateToProps)(EditProblemView));
```
`injectIntl` probably not needed here. It is mostly used if you need to inject `intl` into props.

https://github.com/open-craft/frontend-lib-content-components/blob/476c450e6c6357e34ddf7c011c6e2237345db430/src/editors/containers/TextEditor/index.jsx#L133-L149

navinkarkera reviewed (approved):
:+1:

- [x] I tested this: Tested saving problems from mfe and reloading.
- [x] I read through the code
- [ ] I checked for accessibility issues
- [ ] Includes documentation
- [ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

navinkarkera committed:
feat: answer widget for problem editor (#3)

Implements answer widget for basic problem types[1] based on the design
displayed in this figma board[2]

The redux state is updated on any change in answer widget like adding an
answer, updating existing option, adding feedback etc. The feedback
component supports displaying selected/unselected feedback boxes or
single feedback box based on parsed markdown.

1. https://github.com/edx/frontend-lib-content-components/blob/ab50490cb35ff7c114a15c2a687b5935d380105d/src/editors/data/constants/problem.js#L11
2. https://www.figma.com/file/eIYzYuPvyaFn5IijR3dFe3/Project-Board%3A-Content-Libraries-and-Editing?node-id=1130%3A25243
